server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -1036,27 +1007,29 @@ private void checkForTimeout(long requestId) {
 
     @Override
     public void onConnectionClosed(Transport.Connection connection) {
-        try {
-            List<Transport.ResponseContext<? extends TransportResponse>> pruned =
+        List<Transport.ResponseContext<? extends TransportResponse>> pruned =
                 responseHandlers.prune(h -> h.connection().getCacheKey().equals(connection.getCacheKey()));
+        if (pruned.isEmpty() == false) {
```

Comment:
Mostly we don't have any open handlers here on close, no point forking off for an empty list.